client/service/src/client/wasm_overwrite.rs

```diff
@@ -0,0 +1,179 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+use std::{
+	fs, collections::{HashMap, hash_map::DefaultHasher}, path::Path,
+	hash::Hasher as _,
+};
+use sp_core::traits::FetchRuntimeCode;
+use sp_state_machine::BasicExternalities;
+use sp_blockchain::Result;
+use sc_executor::RuntimeInfo;
+use sp_version::RuntimeVersion;
+use sp_core::traits::RuntimeCode;
+
+#[derive(Clone, Debug)]
+/// Auxiliary structure that holds a wasm blob and its hash.
+struct WasmBlob {
+	code: Vec<u8>,
+	hash: Vec<u8>,
+}
+
+impl WasmBlob {
+	fn new(code: Vec<u8>) -> Self {
+		let hash = make_hash(&code);
+		Self { code, hash }
+	}
+
+	fn runtime_code(&self, heap_pages: Option<u64>) -> RuntimeCode {
+		RuntimeCode {
+			code_fetcher: self,
+			hash: self.hash.clone(),
+			heap_pages,
+		}
+	}
+}
+
+/// Make a hash out of a byte string using the default rust hasher
+fn make_hash<K: std::hash::Hash + ?Sized>(val: &K) -> Vec<u8> {
+	let mut state = DefaultHasher::new();
+	val.hash(&mut state);
+	state.finish().to_le_bytes().to_vec()
+}
+
+impl FetchRuntimeCode for WasmBlob {
+	fn fetch_runtime_code<'a>(&'a self) -> Option<std::borrow::Cow<'a, [u8]>> {
+		Some(self.code.as_slice().into())
+	}
+}
+
+
+/// Scrapes WASM from a folder
+/// and returns WASM from that folder if the runtime spec version matches.
```

Comment:
We should document that the code will read **all** files in the given path and try to instantiate `RuntimeVersion`s for each them. If someone sticks a, dunno, `README` in there, it will not work. 
Alternatively we change the code to only scan for `.wasmblob` files or whatever extension we use (and document that too). I for one would prefer this.

Reply:
Added this with the `.wasm` extension (since that's what's used for releases currently)

However, I think there's an even better solution which is passing each file through the WASM `::validate()` function, which checks if a vector of bytes is a valid wasm executable: https://substrate.dev/rustdocs/v2.0.0/js_sys/WebAssembly/fn.validate.html. This way, the extension shouldn't matter as long as the file is valid wasm.

I'll leave that for another PR however, since it might require some changes/refactoring

Reply:
Filtering on `.wasm` is fine imo. Simple, efficient. Ty.